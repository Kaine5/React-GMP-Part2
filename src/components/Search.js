import {
  useParams,
  useSearchParams,
  useNavigate,
} from "react-router-dom";

const Search = () => {
  const params = useParams();
  const [searchParams, setSearchParams] = useSearchParams();
  const navigate = useNavigate();

  const genre = searchParams.get("genre") || "";
  const sortBy = searchParams.get("sortBy") || "";
  const movieId = searchParams.get("movieId") || "";

  const onSearchSubmit = (e) => {
    e.preventDefault();
    navigate({
      pathname: `${e.target[0].value}`,
      search: `${searchParams}`
    })
  };

  const handleSelectChange = (e, type) => {
    let newValue = e.target.value;
    if (newValue === "") {
      searchParams.delete(type);
    } else {
      searchParams.set(type, e.target.value);
    }
    setSearchParams(searchParams);
  };

  return (
    <div>
      <form onSubmit={onSearchSubmit}>
        <input type="text" defaultValue={params.searchQuery} />
      </form>
      <select value={genre} onChange={(e) => handleSelectChange(e, "genre")}>
        <option value="">Genre</option>
        <option value="1">1</option>
        <option value="2">2</option>
        <option value="3">3</option>
        <option value="4">4</option>
        <option value="5">5</option>
        <option value="6">6</option>
        <option value="7">7</option>
        <option value="8">8</option>
      </select>
      <select value={sortBy} onChange={(e) => handleSelectChange(e, "sortBy")}>
        <option value="">Sort By</option>
        <option value="Release Date">Release Date</option>
        <option value="Name">Name</option>
      </select>
    </div>
  );
};

export default Search;
