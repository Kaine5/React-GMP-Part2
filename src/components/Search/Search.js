import { useMemo } from "react";
import { useParams, useSearchParams, useNavigate } from "react-router-dom";
import useFetching from "../../hooks/useFetching";

const URL = "https://6370ee800399d1995d86faeb.mockapi.io/api/movies";

const Search = () => {
  const params = useParams();
  const [searchParams, setSearchParams] = useSearchParams();
  const navigate = useNavigate();

  const URLForFetching = useMemo(() => {
    let newURL = `${URL}?${
      params.searchQuery ? `name=${params.searchQuery}&` : ""
    }${searchParams.toString()}`;
    return newURL;
  }, [params, searchParams]);

  const [data, loading, error] = useFetching(`${URLForFetching}`);

  const genre = searchParams.get("genre") || "";
  const sortBy = searchParams.get("sortBy") || "";

  const onSearchSubmit = (e) => {
    e.preventDefault();
    navigate({
      pathname: `${e.target[0].value}`,
      search: `${searchParams}`,
    });
  };

  const handleSelectChange = (e, type) => {
    let newValue = e.target.value;
    if (newValue === "") {
      searchParams.delete(type);
    } else {
      searchParams.set(type, e.target.value);
    }
    setSearchParams(searchParams);
  };

  return (
    <div>
      <div>
        <form onSubmit={onSearchSubmit} data-cy="form">
          <input
            type="text"
            defaultValue={params.searchQuery}
            title="searchInput"
            data-cy="searchInput"
          />
        </form>
        <select
          value={genre}
          onChange={(e) => handleSelectChange(e, "genre")}
          title="genreSelect"
          data-cy="genreSelect"
        >
          <option value="">Genre</option>
          <option value="Electronic">Electronic</option>
          <option value="Rap">Rap</option>
          <option value="Hip Hop">Hip Hop</option>
          <option value="Funk">Funk</option>
          <option value="Metal">Metal</option>
          <option value="Reggae">Reggae</option>
          <option value="Country">Country</option>
          <option value="Soul">Soul</option>
        </select>
        <select
          value={sortBy}
          onChange={(e) => handleSelectChange(e, "sortBy")}
          title="sortSelect"
          data-cy="sortSelect"
        >
          <option value="">Sort By</option>
          <option value="releaseDate">Release Date</option>
          <option value="name">Name</option>
        </select>
      </div>
      <div data-cy="searchResult">
        {loading ? (
          <div>Loading</div>
        ) : error ? (
          <div> An error has occured </div>
        ) : (
          data.map((movie) => (
            <div key={movie.id}>
              <div>{movie.name}</div>
              <div>{movie.genre}</div>
              <div>{movie.releaseDate}</div>
            </div>
          ))
        )}
      </div>
    </div>
  );
};

export default Search;
