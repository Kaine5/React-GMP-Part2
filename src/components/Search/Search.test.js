import React from 'react';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import Search from './Search';
import userEvent from '@testing-library/user-event';

it('renders correctly according to the snapshot', () => {
    const { asFragment } = render(<MemoryRouter><Search /></MemoryRouter>);
    expect(asFragment()).toMatchSnapshot();
})

it('allows user to input in the search field', () => {
    render(<MemoryRouter><Search /></MemoryRouter>);

    userEvent.type(screen.getByTitle('searchInput'), "something");
    expect(screen.getByTitle('searchInput')).toHaveValue("something");
    
})