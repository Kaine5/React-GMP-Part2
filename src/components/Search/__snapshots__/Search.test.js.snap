// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly according to the snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form>
        <input
          title="searchInput"
          type="text"
          value=""
        />
      </form>
      <select>
        <option
          value=""
        >
          Genre
        </option>
        <option
          value="Electronic"
        >
          Electronic
        </option>
        <option
          value="Rap"
        >
          Rap
        </option>
        <option
          value="Hip Hop"
        >
          Hip Hop
        </option>
        <option
          value="Funk"
        >
          Funk
        </option>
        <option
          value="Metal"
        >
          Metal
        </option>
        <option
          value="Reggae"
        >
          Reggae
        </option>
        <option
          value="Country"
        >
          Country
        </option>
        <option
          value="Soul"
        >
          Soul
        </option>
      </select>
      <select>
        <option
          value=""
        >
          Sort By
        </option>
        <option
          value="releaseDate"
        >
          Release Date
        </option>
        <option
          value="name"
        >
          Name
        </option>
      </select>
    </div>
    <div>
      <div>
        Loading
      </div>
    </div>
  </div>
</DocumentFragment>
`;
